adjoint-io/fcl
# Reworking soundness check for workflows + QuickCheck tests

More efficient soundness checking
=================================

This pull request is ought to improve the efficiency of the soundness checking algorithm for FCL scripts. The design docs for this change can be seen [here](https://github.com/adjoint-io/design-docs/blob/master/FCL/Soundness.md).

Free choice Petri nets
----------------------

Soundness checking for general workflow nets is infeasible or really complicated. Fortunately, we can express all of our current workflows using free choice Petri nets. Free choice Petri nets are a subclass of Petri nets. They capture the property of "independent execution" of AND branches. This means that the execution of and AND branch cannot be impacted by any other parts of the net.

Free choice Petri net soundness checking is much easier and more efficient.

Sound workflow generation
-------------------------

To have more faith in the correctness of the soundness checking algorithm, we need QuickCheck tests. However, generating arbitrary workflows would yield no helpful information, since we do not know whether they are sound, or if they are unsound, where the problem is. This means we wouldn't be able to verify the correctness of the soundness checking algorithm.

To circumvent this problem, we will generate arbitrary _sound_ workflows. More precisely, we will introduce certain constructors from which workflow nets can be built. The constructors will be really simple so that it is easy to prove their soundness. The net built from these will be sound _by construction_. The last step is to write `Arbitrary` instances of this new data type.

Questions
------------

Generating unsound workflows is still an open question. One idea is to generate a sound workflow, then introduce transitions that make it unsound. This way, we will know where the problem is, so will be able to check for it.

Manual testing
-----------

Load `Language.FCL.GraphvizHacky` into GHCi. Call any of the `generate...NetsSVGs` functions. This will generate the SVGs for the corresponding safe workflow nets into "examples/gen/...". First, you should create those directories yourself:

- `examples/gen/basic`
- `examples/gen/arbitrary`
- `examples/gen/examples`

Status
------

- [x] Haddock strings
- [x] Refactor `WorkflowGen`
- [x] Move `WorkflowGen` to `tests`
- [x] Add instructions on manual testing
- [x] Clean up the unnecessary files, refactor modules (`WorkflowGenExamples`, `WorkflowGenTest`, `GraphvizHacky`)
- [x] Change logs
